hudi-client/src/main/java/org/apache/hudi/metadata/HoodieMetadataIndex.java

```diff
@@ -0,0 +1,189 @@
+    JavaPairRDD<String, String> partitionRecordKeyPairRDD =
+        hoodieKeys.mapToPair(key -> new Tuple2<>(key.getPartitionPath(), key.getRecordKey()));
+
+    // Lookup location of all the hoodie keys
+    JavaPairRDD<HoodieKey, HoodieRecordLocation> recordKeyLocationRDD = lookupIndex(hoodieKeys, jsc, hoodieTable);
+    JavaPairRDD<HoodieKey, String> keyHoodieKeyPairRDD = hoodieKeys.mapToPair(key -> new Tuple2<>(key, null));
+
+    return keyHoodieKeyPairRDD.leftOuterJoin(recordKeyLocationRDD).mapToPair(keyLoc -> {
+      Option<Pair<String, String>> partitionPathFileidPair;
+      if (keyLoc._2._2.isPresent()) {
+        partitionPathFileidPair = Option.of(Pair.of(keyLoc._1().getPartitionPath(), keyLoc._2._2.get().getFileId()));
+      } else {
+        partitionPathFileidPair = Option.empty();
+      }
+      return new Tuple2<>(keyLoc._1, partitionPathFileidPair);
+    });
+  }
+
+  /**
+   * Tag each record with its current location (partition / fileId).
+   *
+   * @param recordRDD   records to tag
+   * @param jsc         spark context
+   * @param hoodieTable hoodie table object
+   */
+  @Override
+  public JavaRDD<HoodieRecord<T>> tagLocation(JavaRDD<HoodieRecord<T>> recordRDD, JavaSparkContext jsc,
+                                              HoodieTable<T> hoodieTable) throws HoodieIndexException {
+    JavaPairRDD<HoodieKey, HoodieRecord> keyRecordPairRDD = recordRDD.mapToPair(r -> new Tuple2<>(r.getKey(), r));
+    JavaPairRDD<HoodieKey, HoodieRecordLocation> keyLocationPairRDD =
+        lookupIndex(keyRecordPairRDD.keys(), jsc, hoodieTable);
+
+    return keyRecordPairRDD.leftOuterJoin(keyLocationPairRDD).values().map(v1 -> {
+      return HoodieIndexUtils.getTaggedRecord(v1._1, Option.ofNullable(v1._2.orNull()));
+    });
+  }
+
+  /**
+   * Update the index (if required) from the results of the writes into the hoodie table.
+   *
+   * @param writeStatusRDD  status of the writes
+   * @param jsc             spark context
+   * @param hoodieTable     hoodie table object
+   */
+  @Override
+  public JavaRDD<WriteStatus> updateLocation(JavaRDD<WriteStatus> writeStatusRDD, JavaSparkContext jsc,
+                                             HoodieTable<T> hoodieTable)
+      throws HoodieIndexException {
+    // No update required for metadata index
+    return writeStatusRDD;
+  }
+
+  /**
+   * Lookup the location for each hoodie key and return the pair<key, location> for all record keys already
+   * present and drop the record keys if not present.
+   */
+  private JavaPairRDD<HoodieKey, HoodieRecordLocation> lookupIndex(JavaRDD<HoodieKey> hoodieKeys,
+                                                                   JavaSparkContext jsc,
+                                                                   HoodieTable<T> hoodieTable) {
+    SliceView fsView = hoodieTable.getSliceView();
+    List<HoodieBaseFile> baseFiles = fsView.getLatestFileSlices(HoodieMetadataImpl.METADATA_PARTITION_NAME)
```

Comment:
this has to be derived from the input keys eventually, to support updates to different metadata partitions?

Reply:
Yes. Since we dont have other metadata table partitions, I have kept it simple.